### PR TITLE
Mustache SuperCluster Regression for 800

### PR DESCRIPTION
**Pull request by matteosan1**
The PR updates the online and offline regression for mustache SC. 
At the moment the corrections are off by default due to the useRegression flag set to false. 
Once activated:
- Offline regression impact only pixel-matching since it is the only place where the energy is taken directly from the mustache supercluster.
- Online regression represent the "highest" level energy corrections used at HLT for electrons and photons.
  We plan to switch the corrections on once the payloads will be integrated into the DB.

### CODE REVIEW

File: RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc [hunk 1]
```diff
@@ -0,0 +1,316 @@
+//#include <TFile.h>
+#include "../interface/SCEnergyCorrectorSemiParm.h"
```
Comment (slava77) [inline RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc]: relative paths are not allowed

File: RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc [hunk 2]
```diff
@@ -0,0 +1,316 @@
+//#include <TFile.h>
```
Comment (slava77) [inline RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc]: please drop the commented out TFile include

File: RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+#include "CondFormats/EgammaObjects/interface/GBRForestD.h"
+#include "CondFormats/DataRecord/interface/GBRDWrapperRcd.h"
+#include "FWCore/Framework/interface/ESHandle.h" 
+#include "DataFormats/EgammaCandidates/interface/Photon.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterTools.h"
+#include "DataFormats/EcalDetId/interface/EcalSubdetector.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "RooWorkspace.h"
```
Comment (slava77) [inline RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc]: I think all Roo\* includes are not needed. Please check and clarify if it is not the case.

File: RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    eval[17] = clusterMaxDRRawEnergy/raw_energy; //clusterMaxDRRawEnergy/scRawEnergy
+    eval[18] = clusterRawEnergy[0]/raw_energy; //clusterRawEnergy[0]/scRawEnergy
+    eval[19] = clusterRawEnergy[1]/raw_energy; //clusterRawEnergy[1]/scRawEnergy
+    eval[20] = clusterRawEnergy[2]/raw_energy; //clusterRawEnergy[2]/scRawEnergy
+    eval[21] = clusterDPhiToSeed[0]; //clusterDPhiToSeed[0]
+    eval[22] = clusterDPhiToSeed[1]; //clusterDPhiToSeed[1]
+    eval[23] = clusterDPhiToSeed[2]; //clusterDPhiToSeed[2]
+    eval[24] = clusterDEtaToSeed[0]; //clusterDEtaToSeed[0]
+    eval[25] = clusterDEtaToSeed[1]; //clusterDEtaToSeed[1]
+    eval[26] = clusterDEtaToSeed[2]; //clusterDEtaToSeed[2]
```
Comment (slava77) [inline RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc]: almost all of the commented out text is not a comment and looks more like an automated repeat of the active code.
Please explain why this is needed.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc [hunk 5]
```diff
@@ -0,0 +1,307 @@
+#include "RecoEgamma/EgammaTools/interface/SCEnergyCorrectorSemiParm.h"
+#include "CondFormats/DataRecord/interface/GBRDWrapperRcd.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterTools.h"
+#include "DataFormats/EcalDetId/interface/EcalSubdetector.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "TStreamerInfo.h"
```
Comment (slava77) [inline RecoEgamma/EgammaTools/src/SCEnergyCorrectorSemiParm.cc]: TStreamerInfo looks unnecessary as well

**Comment by slava77** [comment]
.. the IB used for the tests is broken. So, the tests have failed.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@matteosan1 
I misunderstood the impact of having the regression disabled.
I was expecting that this is going to be a technical PR with no change to the physics outputs with the flag "false" until the new regression is available.
If I understand correctly, disabled regression correction is making the performance worse. 
If so, this PR can not go to the release until there is an update to enable the regression.

Is it possible to keep the old training/regression in place and just swap out when the new one is available?

just to throw in an example from the comparisons for 134.911:
all_oldvsnew_runsingleph2015dwf134p911c_log10recocaloclusters_particleflowsuperclusterecal_particleflowbasicclusterecalbarrel_rereco_obj_energy
all_oldvsnew_runsingleph2015dwf134p911c_log10recocaloclusters_particleflowsuperclusterecal_particleflowbasicclusterecalendcap_rereco_obj_energy
impact on PF candidates is rather small in general, but differences are present for both ged electrons and photons (small reduction in the count; can't say more certainly with the small set of tests in jenkins).

**Comment by matteosan1** [comment]
Hi Slava,
I am really puzzled since the regression on or off cannot impact the PF
basic clusters because it is applied on top of the SuperCluster only.

Regards,
     Matteo.

+++++++++++
Matteo Sani
University of California, San Diego, US.
Tel: +41227671577
CERN Office: 40-3-A02

**Comment by slava77** [comment]
Hi Matteo,

I judged purely by the values that come out in the saved objects (an example plot posted earlier).
I didn't look at the code where the values are changed or set.
Maybe the updates affected something unexpected.

**Comment by matteosan1** [comment]
Hi Slava,
investigating a little bit and talking to Lindsey I realized that last year
we were actually using the old 70X regression and not "nothing" as I was
expecting that's why you see some degradation.
So I'll take care of the payload and once they are in the DB I'll update
the PR switching on the corrections.

Regards,
     Matteo.

+++++++++++
Matteo Sani
University of California, San Diego, US.
Tel: +41227671577
CERN Office: 40-3-A02

**Comment by mmusich** [comment]
@matteosan1 please let us know when we should integrate new conditons in GT. @franzoni you might want to follow this as well.

**Comment by matteosan1** [comment]
Yes I am working on that right now.

Regards,
     Matteo.

+++++++++++
Matteo Sani
University of California, San Diego, US.
Tel: +41227671577
CERN Office: 40-3-A02

**Comment by slava77** [comment]
@matteosan1 
please update on the status

**Comment by slava77** [comment]
:clock11:

**Comment by Martin-Grunewald** [comment]
what's the progress (or lack thereof) here?

**Comment by matteosan1** [comment]
Hi Martin,
we are waiting for the integration of the regression payloads into the GT.
I have the final version of the code ready.
@mmusich

Regards,
     Matteo.

+++++++++++
Matteo Sani
University of California, San Diego, US.
Tel: +41227671577
CERN Office: 40-3-A02

**Comment by mmusich** [comment]
@matteosan1 I haven't seen the announcement of new payloads, though I see now the tags are in queue. Please announce when tags are ready on the AlCa HN forum.

**Comment by mmusich** [comment]
@matteosan1 @slava77  
Updated Global Tags are here: https://github.com/cms-sw/cmssw/pull/13141
To get them:
```
git remote add diguida https://github.com/mmusich/cmssw.git
git fecth mmusich
git cherry-pick 5f9cf6a4e701f3e1553c76f60e52ba59c0fbaf1d
```

@Martin-Grunewald N.B.: no changes yet for online Global Tag (pending payloads from @matteosan1)

**Comment by matteosan1** [comment]
Two validation plots showing the new regression on Z->ee events (blue is the new one, red the current default)
mustache_offline_regression_ee
mustache_offline_regression_eb

.

**Comment by slava77** [comment]
this PR needs a rebase.
It looks like instead of cherry-picking the change in autoCond.py we get a merge of all recent 80X changes

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mmusich** [comment]
+1

**Comment by slava77** [comment]
@mmusich I'm hoping that there is no more update in the autoCond.py planned imminently.
If that changes, this PR will hopefully become unmergeable, but can instead override changes.
So, @matteosan1,   it may be better to remove the commit changing autoCond

**Comment by mmusich** [comment]
@slava77 not imminent, but soon we will have to add few ingredients for the final definition 80X asymptotic scenarios, i.e. Ecal noise, Ecal Selective readout, and a bunch of other ites, so yes I would be definitely better to remove the commit

**Comment by slava77** [comment]
@matteosan1 which energies did you plot in https://github.com/cms-sw/cmssw/pull/13000#issuecomment-177917079 ?
Are these cluster level or the final candidate (e.g. gedGsfElectron) pt or energy()?

Most of the significant changes that I observe at the high level object level (gedGsfElectron, gedPhoton, and particleFlow candidates) are all localized to pt~10 GeV which seems to be just a cluster threshold effects:
e.g. gamma pt=10 GeV gun events:
all_sign665vsorig_singlegamma13pt10wf1318p0c_recophotons_gedphotons__reco_obj_et

vs gamma 35 GeV sample
all_sign665vsorig_singlegamma13pt35wf1319p0c_recophotons_gedphotons__reco_obj_et

Similar story for gsfElectrons and PFCandidates (particleFlow) with electron or photon ID.

DQM doesn't seem to have much to offer in terms of ECAL SC reco/true energy plots.
This one in electrons (in ele 35 GeV gun sample) appears to show an improvement of a similar kind posted by Matteo
wf1317_elescovertrue_pf_vs_egamma

At the lower level, the supercluster energies are changing noticeably: plots from H->gammagamma (workflow 1332, no pileup)
wf1332_ebsc_e
wf1332_eesc_e



There is a rather significant change in R9 plots (but I suspect these are plots of an incorrect variable; if the actual "good" R9 is changing: it looks like more clusters are pushed to lower R9 values, which is not so good)
here is a plot in H->gammagamma
wf1332_ebsc_r9
wf1332_eesc_r9

Based on the PR description, the effects appear to be somewhat expected, but some comments are needed to confirm the behavior is expected before this can be signed off.

**Comment by matteosan1** [comment]
Hi Slava, 
SC energy plots look consistent to mine, I am a bit surprised of R9 and I am running a simple test to cross-check it.

**Comment by slava77** [comment]
On 2/3/16 5:02 AM, Matteo Sani wrote:
> Hi Slava,
> SC energy plots look consistent to mine, I am a bit surprised of R9 and
> I am running a simple test to cross-check it.
> 

I recall that in the past we had R9 plot in DQM which was not computed
correctly (using inconsistent numerator and denominator).
It may be worth to check the code that makes the plot (it should be easy
given the path and the histogram name).
Someone in ECAL/EGM should fix it, if it is the case here.

**Comment by matteosan1** [comment]
Hi Slava, 
I don't see big issues with R9 in my plots (attached). I'll follow up on the DQM issue with ECAL guys.
ele_r9_eb
ele_r9_ee

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/13000	09e078a
- code changes are in line with the description (a different SC correction is applied)
- jenkins test pass and comparisons with baseline show differences consistent with the local tests summarized earlier in https://github.com/cms-sw/cmssw/pull/13000#issuecomment-178765103

@davidlange6 
while it's still coherent, please merge this at your earliest.
If there are some issues and this can't be merged imminently, this PR will have to be updated to remove changes in autoCond.py file https://github.com/cms-sw/cmssw/pull/13000#issuecomment-178673982

**Comment by davidlange6** [comment]
+1

**Pull request merged**